### PR TITLE
Development

### PR DESCRIPTION
**Pull request by pedroartsimao**
Code test finished. Waiting for a review :+1:

### CODE REVIEW

File: app/src/main/java/com/arctouch/codetest/presenter/PresenterHolder.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.arctouch.codetest.presenter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class PresenterHolder {
+    static volatile PresenterHolder singleton = null;
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/presenter/PresenterHolder.java]: Why are you using volatile here? Is this class being accessed from different threads?

File: app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    private List<Stop> stopList = new ArrayList<>();
+    private List<Departure> departureList = new ArrayList<>();
+
+    public RouteItemPresenterImpl(RouteItemView view) {
+        this.view = view;
+    }
+
+    @Override
+    public void create(Route route) {
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java]: What is being created here?

File: app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                    @Override
+                    public void onFailure(Call<FindStopsByRouteIdResponse> call, Throwable t) {
+                        t.printStackTrace();
+                    }
+                });
+    }
+
+    @Override
+    public void loadDepartures() {
+        Params params = new Params();
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java]: How could you remove the duplication here?

File: app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+import com.arctouch.codetest.ui.listdecoration.DividerItemDecoration;
+import com.arctouch.codetest.view.RouteListView;
+
+import java.util.List;
+
+public class MainActivity extends AppCompatActivity implements RouteListView, SearchView.OnQueryTextListener, RoutesAdapter.OnItemClickListener {
+
+    private RouteListPresenter presenter;
+
+    private boolean mTwoPane;
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java]: Why are you using the m prefix for this variable? What do you think about this kind of convention?

File: app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    private void setupViews() {
+        progressBar = (ProgressBar) findViewById(R.id.progress_bar);
+        tvPlaceholder = (TextView) findViewById(R.id.tv_placeholder);
+    }
+
+    @Override
+    public boolean onQueryTextSubmit(String query) {
+        if(query.isEmpty())
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java]: Why are you omitting braces here?

File: app/src/main/java/com/arctouch/codetest/ui/fragment/ItemDetailFragment.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.List;
+
+
+public class ItemDetailFragment extends Fragment implements RouteItemView {
+
+    public static final String ARG_ITEM_OBJ = "item_obj";
+
+    private RouteItemPresenter presenter;
+
+    private ProgressBar pbStops, pbDepartures;
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/ui/fragment/ItemDetailFragment.java]: Why are you abbreviating these variable names?

File: app/src/test/java/com/arctouch/codetest/ExampleUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package com.arctouch.codetest;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * To work on unit tests, switch the Test Artifact in the Build Variants view.
+ */
+public class ExampleUnitTest {
```
Comment (arctouch-andrecardoso) [inline app/src/test/java/com/arctouch/codetest/ExampleUnitTest.java]: What would you use to write proper unit tests? Why you have chosen to not write them?

File: app/src/main/java/com/arctouch/codetest/presenter/Presenter.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.arctouch.codetest.presenter;
+
+public interface Presenter {
```
Comment (arctouch-andrecardoso) [inline app/src/main/java/com/arctouch/codetest/presenter/Presenter.java]: Why did you choose this MVP approach? What are the main advantages of it?

**Comment by arctouch-andrecardoso** [comment]
Hi @pedroartsimao,

I have finished my review. Please take a look at my comments. 

Good luck!

File: app/src/main/java/com/arctouch/codetest/presenter/PresenterHolder.java (hunk 1 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/presenter/PresenterHolder.java]: Yes, it's a possibility. This feature is not used in this project, but I'm used to make it volatile by default.

File: app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java (hunk 2 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java]: It's a reference for onCreate, where I call this method for setting up the data.

File: app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java (hunk 3 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/presenter/implementations/RouteItemPresenterImpl.java]: I could prepare a App Glu API object once and implements Callback. But I actually would do it in another abstraction layer of data access.

File: app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java (hunk 4 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java]: It's a simple variable generated by Android Studio when I started the project and I forgot to change it. For real development m prefix must not be used, it's not clear enough in complex codes. In sample codes it helps to identify simple variables that you creating or referring.

File: app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java (hunk 5 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/ui/activity/MainActivity.java]: Fail! I made a mistake. It's not a good practice.

File: app/src/main/java/com/arctouch/codetest/ui/fragment/ItemDetailFragment.java (hunk 6 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/ui/fragment/ItemDetailFragment.java]: It's a simple way to refer to a view item and know it type. TextView, ProgressBar, EditText, etc... People where I'm working nowadays are using this kind of abbreviation and I'm used to it.

File: app/src/test/java/com/arctouch/codetest/ExampleUnitTest.java (hunk 7 above)
Comment (pedroartsimao) [inline app/src/test/java/com/arctouch/codetest/ExampleUnitTest.java]: JUnit with Mockito. I'm not automated unit test expert and I've worked and studied a lot these last weeks and I didn't prioritized unit tests for this project.

File: app/src/main/java/com/arctouch/codetest/presenter/Presenter.java (hunk 8 above)
Comment (pedroartsimao) [inline app/src/main/java/com/arctouch/codetest/presenter/Presenter.java]: MVP has a great balance between complexity and benefits. It helps to remove some rules from Activity (View), Mock objects in automated tests, create a layer of data access abstraction using repository pattern. I made a simple implementation of MVP (based on Israel Camacho's Effective Android project) for this project trying to keep it simple and organized.

**Pull request merged**